Reader and writer stalled with small-key workload in longevity run

**Observed that both writer and reader got stalled during longevity run workload  `small-key`  with `JDK 11` Image on client side after `~20h 14m`** 

Note: No reader, writer  failure observed during this run.

Environment details: PKS / K8 with medium cluster:
```
3 master: xlarge: 4 CPU, 16 GB Ram, 32 GB Disk
5 worker: 2xlarge: 8 CPU, 32 GB Ram, 64 GB Disk
Tier-1 storage is from VSAN datastore
Tier-2 storage curved on NFS Client provisioner using Isilon as backend
```
```
Pravega version : 0.5.0-2303.7b5f02f
Zookeeper Operator : pravega/zookeeper-operator:0.2.2
Pravega Operator: pravega/pravega-operator:0.3.3-rc1
```
```
INFO  [2019-07-23 02:53:03,075] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 20h 12m
INFO  [2019-07-23 02:53:03,075] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:159,143,987, events/sec:2,187, KB/sec:1736.8623

INFO  [2019-07-23 02:53:03,076] io.pravega.longevity.utils.PerformanceUtils: Readers (6/6): events:159,129,973, events/sec:2,187, KB/sec:1736.709

```

```
INFO  [2019-07-23 02:55:22,570] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 20h 14m
INFO  [2019-07-23 02:55:22,570] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:159,143,987, events/sec:2,183, KB/sec:1733.8594

INFO  [2019-07-23 02:55:22,570] io.pravega.longevity.utils.PerformanceUtils: Readers (6/6): events:159,129,973, events/sec:2,183, KB/sec:1733.707

```
As per discussion seems like this issue is not similar to #3900 as there is no abrupt stall and more likely test stops writing events if more than 128 writer futures does not complete.
The writes stops @ around `2019-07-23 02:53:03`

```INFO  [2019-07-23 02:52:21,340] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:159,143,985, events/sec:2,188, KB/sec:1737.8408
INFO  [2019-07-23 02:53:03,075] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:159,143,987, events/sec:2,187, KB/sec:1736.8623
```

The `small-key` test stops writing events if more than 128 writer futures does not complete. ( the write future is returned by `io.pravega.client.stream.impl.EventStreamWriterImpl#writeEvent(...)` )
   
```"throttle": {
        "maxOutstandingAcks": 128
      },
```

Note: Have shared the logs in internal slack separately.

Updates for the logs shared on 25/Jul.

*Summary* : The small-key worker pod observes a java.lang.OutOfMemoryError causing the reads and writers to stop working. Also the worker pod is not able to send stats
to the manager pod due to `javax.ws.rs.ProcessingException`

*Details:* 
[1]
Observing the writers stop writing around the time stamp 9:23 on the small key manager.
```
// snip at 09:23
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 20h 20m
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:167,457,691, events/sec:2,286, KB/sec:1815.8652
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 6,328,437, average block time (ms): 18
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: Readers (6/6): events:167,376,934, events/sec:2,285, KB/sec:1814.9902
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: New runtime for 1c90127e-3be8-4f7e-b3f6-1a3d19d2cf18  Running for 20h 20m
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: Dumping stats for stream longevity/smallKey
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:167,457,691, events/sec:2,287, KB/sec:1816.6348
INFO  [2019-07-25 09:23:10,272] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0

//snip @ 09:37...

INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Got new data  Running for 20h 34m
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:167,457,691, events/sec:2,260, KB/sec:1795.2969
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 6,328,437, average block time (ms): 18
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Readers (6/6): events:167,378,644, events/sec:2,259, KB/sec:1794.4492
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: New runtime for 1c90127e-3be8-4f7e-b3f6-1a3d19d2cf18  Running for 20h 34m
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Dumping stats for stream longevity/smallKey
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Writers (3/3): events:167,457,691, events/sec:2,261, KB/sec:1796.0244
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils:  Transactions: started: 0, perSec:0.0, committed:0, failed:0, aborted:0, aborted but read: 0
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils:  Blocked Waiting For ACK: times blocks: 6,328,437, average block time (ms): 18
INFO  [2019-07-25 09:37:09,121] io.pravega.longevity.utils.PerformanceUtils: Readers (6/6): events:167,378,644, events/sec:2,260, KB/sec:1795.1768
```

[2]  Snip which indicates the readers and writers due to OutOfMemoryError
```
2019-07-25 09:12:22,889 72590373 [epollEventLoopGroup-4-2] DEBUG i.p.client.netty.impl.FlowHandler - Exception observed for flow id 17 due to Java heap space

2019-07-25 09:12:24,350 72591834 [epollEventLoopGroup-4-2] INFO  i.p.c.s.i.SegmentOutputStreamImpl - Handling exception io.pravega.shared.protocol.netty.ConnectionFailedException: java.lang.OutOfMemoryError: Java heap space for connection io.pravega.client.netty.impl.ClientConnectionImpl@407ffe79 on writer e5994950-371d-4e57-8a6b-499ac1f8fcee. SetupCompleted: true, Closed: false

2019-07-25 09:12:24,351 72591835 [epollEventLoopGroup-4-2] WARN  i.p.c.s.i.AsyncSegmentInputStreamImpl - Processing failure:
io.pravega.shared.protocol.netty.ConnectionFailedException: java.lang.OutOfMemoryError: Java heap space
```

[3] The worker pod is also not able to send stats to the longevity manager.
```
2019-07-25 09:14:09,775 72697259 [pool-2-thread-1] ERROR i.p.l.t.w.TaskPerformanceCollector - Error sending Performance Stats
javax.ws.rs.ProcessingException: Already connected
        at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:261)
        at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:684)
        at org.glassfish.jersey.client.JerseyInvocation$1.call(JerseyInvocation.java:681)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
        at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:444)
        at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:681)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:437)
        at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:343)
        at io.pravega.longevity.utils.restclient.RestClient.post(RestClient.java:109)
        at io.pravega.longevity.testworker.MessageClient.postPerformance(MessageClient.java:32)
        at io.pravega.longevity.testworker.workers.TaskPerformanceCollector.lambda$sendPerformance$14(TaskPerformanceCollector.java:182)
        at io.pravega.longevity.testworker.workers.TaskPerformanceCollector.inLock(TaskPerformanceCollector.java:191)
        at io.pravega.longevity.testworker.workers.TaskPerformanceCollector.sendPerformance(TaskPerformanceCollector.java:171)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:305)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.lang.IllegalStateException: Already connected
        at java.base/sun.net.www.protocol.http.HttpURLConnection.setRequestProperty(HttpURLConnection.java:3155)
        at org.glassfish.jersey.client.internal.HttpUrlConnector.setOutboundHeaders(HttpUrlConnector.java:426)
```

--- a/client/src/main/java/io/pravega/client/segment/impl/EventSegmentReader.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/EventSegmentReader.java
@@ -10,6 +10,7 @@
 package io.pravega.client.segment.impl;
 
 import io.pravega.client.stream.EventStreamWriter;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
 import java.util.concurrent.CompletableFuture;
 
@@ -76,7 +77,7 @@ public interface EventSegmentReader extends AutoCloseable {
      * 
      * @return A future that completes when the request to fill the buffer has returned.
      */
-    public abstract CompletableFuture<Void> fillBuffer();
+    public abstract CompletableFuture<SegmentRead> fillBuffer();
     
     /**
      * Closes this reader. No further methods may be called after close.

--- a/client/src/main/java/io/pravega/client/segment/impl/EventSegmentReaderImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/EventSegmentReaderImpl.java
@@ -14,6 +14,7 @@ import io.pravega.common.LoggerHelpers;
 import io.pravega.shared.protocol.netty.InvalidMessageException;
 import io.pravega.shared.protocol.netty.WireCommandType;
 import io.pravega.shared.protocol.netty.WireCommands;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
 import java.util.concurrent.CompletableFuture;
 import javax.annotation.concurrent.GuardedBy;
@@ -99,7 +100,7 @@ class EventSegmentReaderImpl implements EventSegmentReader {
 
     @Override
     @Synchronized
-    public CompletableFuture<Void> fillBuffer() {
+    public CompletableFuture<SegmentRead> fillBuffer() {
         return in.fillBuffer();
     }
     

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStream.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStream.java
@@ -9,6 +9,7 @@
  */
 package io.pravega.client.segment.impl;
 
+import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
 import java.util.concurrent.CompletableFuture;
 
@@ -62,7 +63,7 @@ public interface SegmentInputStream extends AutoCloseable {
      * 
      * @return A future that will be completed when there is data available to read.
      */
-    public abstract CompletableFuture<Void> fillBuffer();
+    public abstract CompletableFuture<SegmentRead> fillBuffer();
     
     /**
      * Closes this InputStream. No further methods may be called after close.

--- a/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
+++ b/client/src/main/java/io/pravega/client/segment/impl/SegmentInputStreamImpl.java
@@ -10,7 +10,6 @@
 package io.pravega.client.segment.impl;
 
 import com.google.common.base.Preconditions;
-import com.google.common.util.concurrent.Runnables;
 import io.pravega.common.Exceptions;
 import io.pravega.common.concurrent.Futures;
 import io.pravega.common.util.CircularBuffer;
@@ -205,7 +204,7 @@ class SegmentInputStreamImpl implements SegmentInputStream {
 
     @Override
     @Synchronized
-    public CompletableFuture<Void> fillBuffer() {
+    public CompletableFuture<SegmentRead> fillBuffer() {
         log.trace("Filling buffer {}", this);
         Exceptions.checkNotClosed(asyncInput.isClosed(), this);
         try {      
@@ -217,7 +216,7 @@ class SegmentInputStreamImpl implements SegmentInputStream {
             log.warn("Encountered exception filling buffer", e);
             return CompletableFuture.completedFuture(null);
         }
-        return outstandingRequest == null ? CompletableFuture.completedFuture(null) : outstandingRequest.thenRun(Runnables.doNothing());
+        return outstandingRequest == null ? CompletableFuture.completedFuture(null) : outstandingRequest;
     }
     
     @Override

--- a/client/src/test/java/io/pravega/client/stream/impl/OrdererTest.java
+++ b/client/src/test/java/io/pravega/client/stream/impl/OrdererTest.java
@@ -13,6 +13,7 @@ import com.google.common.collect.ImmutableList.Builder;
 import io.pravega.client.segment.impl.EndOfSegmentException;
 import io.pravega.client.segment.impl.EventSegmentReader;
 import io.pravega.client.segment.impl.Segment;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
 import java.util.List;
 import java.util.concurrent.CompletableFuture;
@@ -42,7 +43,7 @@ public class OrdererTest {
         }
 
         @Override
-        public CompletableFuture<Void> fillBuffer() {
+        public CompletableFuture<SegmentRead> fillBuffer() {
             return CompletableFuture.completedFuture(null);
         }
 

--- a/client/src/test/java/io/pravega/client/stream/mock/MockSegmentIoStreams.java
+++ b/client/src/test/java/io/pravega/client/stream/mock/MockSegmentIoStreams.java
@@ -24,6 +24,7 @@ import io.pravega.client.segment.impl.SegmentTruncatedException;
 import io.pravega.client.stream.impl.PendingEvent;
 import io.pravega.common.util.ByteBufferUtils;
 import io.pravega.shared.protocol.netty.WireCommands;
+import io.pravega.shared.protocol.netty.WireCommands.SegmentRead;
 import java.nio.ByteBuffer;
 import java.util.Collections;
 import java.util.List;
@@ -179,7 +180,7 @@ public class MockSegmentIoStreams implements SegmentOutputStream, SegmentInputSt
     }
 
     @Override
-    public CompletableFuture<Void> fillBuffer() {
+    public CompletableFuture<SegmentRead> fillBuffer() {
         return CompletableFuture.completedFuture(null);
     }
 

